extensions/interactions/GraphInput/GraphInput.js

```diff
@@ -187,362 +191,468 @@ oppia.directive('graphViz', function() {
-            edge.src--;
+          if (angular.isDefined(label)) {
+            $scope.graph.vertices[$scope.state.selectedVertex].label = label;
           }
-          if (edge.dst > index) {
-            edge.dst--;
+          return $scope.graph.vertices[$scope.state.selectedVertex].label;
+        };
+        $scope.selectedEdgeWeight = function(weight) {
+          if ($scope.state.selectedEdge === null) {
+            return '';
           }
-          return edge;
-        });
-        $scope.graph.vertices.splice(index, 1);
-        $scope.state.hoveredVertex = null;
-      }
-      $scope.selectedVertexLabelGetterSetter = function(label) {
-        if ($scope.state.selectedVertex === null) {
-          return '';
-        }
-        if (angular.isDefined(label)) {
-          $scope.graph.vertices[$scope.state.selectedVertex].label = label;
+          if (angular.isDefined(weight) && angular.isNumber(weight)) {
+            $scope.graph.edges[$scope.state.selectedEdge].weight = weight;
+          }
+          return $scope.graph.edges[$scope.state.selectedEdge].weight;
+        };
+
+
+        // Styling functions
+        var DELETE_COLOR = 'red';
+        var HOVER_COLOR = 'aqua';
+        var SELECT_COLOR = 'orange';
+        var DEFAULT_COLOR = 'black';
+        $scope.getEdgeColor = function(index) {
+          if ($scope.state.currentMode === _MODES.DELETE &&
+              index === $scope.state.hoveredEdge &&
+              $scope.canDeleteEdge) {
+            return DELETE_COLOR;
+          } else if (index === $scope.state.hoveredEdge) {
+            return HOVER_COLOR;
+          } else if ($scope.state.selectedEdge === index) {
+            return SELECT_COLOR;
+          } else {
+            return DEFAULT_COLOR;
+          }
+        };
+        $scope.getVertexColor = function(index) {
+          if ($scope.state.currentMode === _MODES.DELETE &&
+              index === $scope.state.hoveredVertex &&
+              $scope.canDeleteVertex) {
+            return DELETE_COLOR;
+          } else if (index === $scope.state.currentlyDraggedVertex) {
+            return HOVER_COLOR;
+          } else if (index === $scope.state.hoveredVertex) {
+            return HOVER_COLOR;
+          } else if ($scope.state.selectedVertex === index) {
+            return SELECT_COLOR;
+          } else {
+            return DEFAULT_COLOR;
+          }
+        };
+        $scope.getDirectedEdgeArrowPoints = function(index) {
+          return graphDetailService.getDirectedEdgeArrowPoints(
+            $scope.graph, index);
+        };
+        $scope.getEdgeCentre = function(index) {
+          return graphDetailService.getEdgeCentre($scope.graph, index);
+        };
+      }]
+  };
+});
+
+oppia.factory('graphInputRulesService', [function() {
+  var isIsomophic = function(graph1, graph2) {
+    if (graph1.vertices.length != graph2.vertices.length) {
+      return false;
+    }
+
+    var constructAdjacencyMatrix = function(graph) {
```

Comment:
I think this and nextPermutation can be top-level functions in graphInputRulesService, rather than contained in isIsomorphic.